Content.Server/GameObjects/Components/NodeContainer/Nodes/PipeNode.cs

```diff
@@ -0,0 +1,124 @@
+            serializer.DataField(ref _pipeDirection, "pipeDirection", PipeDirection.None);
+        }
+
+        protected override IEnumerable<Node> GetReachableNodes()
+        {
+            foreach (CardinalDirection direction in Enum.GetValues(typeof(CardinalDirection)))
+            {
+                PipeDirectionFromCardinal(direction, out var ownNeededConnection, out var theirNeededConnection);
+                if ((_pipeDirection & ownNeededConnection) == PipeDirection.None)
+                {
+                    continue;
+                }
+                var pipeNodesInDirection = Owner.GetComponent<SnapGridComponent>()
+                    .GetInDir((Direction) direction)
+                    .Select(entity => entity.TryGetComponent<NodeContainerComponent>(out var container) ? container : null)
+                    .Where(container => container != null)
+                    .SelectMany(container => container.Nodes)
+                    .OfType<PipeNode>()
+                    .Where(pipeNode => (pipeNode._pipeDirection & theirNeededConnection) != PipeDirection.None);
+                foreach (var pipeNode in pipeNodesInDirection)
+                {
+                    yield return pipeNode;
+                }
+            }
+        }
+
+        private void PipeDirectionFromCardinal(CardinalDirection direction, out PipeDirection sameDir, out PipeDirection oppDir)
+        {
+            switch (direction)
+            {
+                case CardinalDirection.North:
+                    sameDir = PipeDirection.Up;
+                    oppDir = PipeDirection.Down;
+                    break;
+                case CardinalDirection.South:
+                    sameDir = PipeDirection.Down;
+                    oppDir = PipeDirection.Up;
+                    break;
+                case CardinalDirection.East:
+                    sameDir = PipeDirection.Right;
+                    oppDir = PipeDirection.Left;
+                    break;
+                case CardinalDirection.West:
+                    sameDir = PipeDirection.Left;
+                    oppDir = PipeDirection.Right;
+                    break;
+                default:
+                    throw new ArgumentException("Invalid Direction.");
+            }
+        }
+
+        private void SetPipeDirection(PipeDirection pipeDirection)
+        {
+            throw new NotImplementedException();
+
+            NodeGroup.RemoveNode(this);
+            ClearNodeGroup();
+            _pipeDirection = pipeDirection;
+            TryAssignGroupIfNeeded();
+            //CombineGroupWithReachable();
+        }
+
+        private enum CardinalDirection
+        {
+            North = Direction.North,
+            South = Direction.South,
+            East = Direction.East,
+            West = Direction.West,
+        }
+    }
+
+    public enum PipeDirection
+    {
+        None = 0,
+
+        //Half of a pipe in a direction
+        Up = 1 << 0,
```

Comment:
I'd be careful with the terms `Up` and `Down`, given that we may eventually have Z-level pipes which would LITERALLY go `Up`/`Down`.

Just go with the standard `North`/`East`/`South`/`West` direction names